ui/ui_v3/scripts/select_table.js

```diff
@@ -0,0 +1,327 @@
+  }
+  // Else detect the table containing the selected range
+  else {
+    dataRange =  selectTable(rowStart,rowEnd,colStart,colEnd,range.getValues(),maxRows,maxCols);
+  }
+    
+  Logger.log('Range of the detected table');
+  Logger.log(dataRange);
+
+  // Find the detected range's a1Notation
+  var rangeTable = sheet.getRange(dataRange[0],dataRange[2],dataRange[1]-dataRange[0]+1,dataRange[3]-dataRange[2]+1);
+  var rangeString = rangeTable.getA1Notation();
+
+  // Select the first row of the detected table as header row
+  var headerRange = sheet.getRange(dataRange[0],dataRange[2],1,dataRange[3]-dataRange[2]+1).getA1Notation();
+  var headerRow = dataRange[0];
+
+  // Storing details about the detected table
+  var dataRange = {
+   'inputSheet': sheet.getName(),
+   'rangeString': rangeString,
+   'headerRow': headerRow,
+   'headerRange': headerRange
+  };
+  var userProperties = PropertiesService.getUserProperties();
+  userProperties.setProperties(dataRange);
+  
+  return dataRange;
+}
+
+/**
+ * Detects the table containing the specified range of cells
+ *
+ * @param {number} rowStart The start row of the specified range
+ * @param {number} rowEnd The end row of the specified range
+ * @param {number} colStart The start column of the specified range
+ * @param {number} colEnd The end column of the specified range
+ * @param {number} maxRows The last row of the sheet containing data
+ * @param {number} maxCols The last column of the sheet containing data
+ * @return {Array} The start row, end row, start column, end column of the detected table
+ */
+function selectTable(rowStart,rowEnd,colStart,colEnd,values,maxRows,maxCols) {
+  /**
+   * Flag is used to indicate if the detected range changes in current function call
+   * Flag is set to true when the range changes
+   * Flag set to false indicates the range did not change
+   */
+  var flag = false;
+  
+  for(var j = colStart; j <= colEnd && j <= maxCols; j++) {
+    // check for blank values in up directorion
+    for(var i = rowStart - 1; i >= 1 && i <= maxRows; i--) {
+      if(values[i-1][j-1] === '') {
+        break;
+      }
+      rowStart = i;
+      flag=true;
+    }
+    // check for blank values in down directorion
+    for(var i = rowEnd + 1; i <= maxRows; i++) {
+      if(values[i-1][j-1] === '') {
+        break;
+      }
+      rowEnd = i;
+      flag=true;
+    }
+  }
+
+  for(var i = rowStart; i <= rowEnd && i <= maxRows; i++) {
+    // check for blank values in left directorion
+    for(var j = colStart - 1; j >= 1 && j <= maxCols ; j--) {
+      if(values[i-1][j-1] === '') {
+        break;
+      }
+      colStart = j;
+      flag=true;
+    }
+    // check for blank values in right directorion
+    for(var j = colEnd + 1; j <= maxCols; j++) {
+      if(values[i-1][j-1] === '') {
+        break;
+      }
+      colEnd = j;
+      flag=true;
+    }
+  }
+
+  // If the detected range is altered the function is called again
```

Comment:
Nit: How does this happen, i.e how is flag modified when there are some changes done to the range. @arushigarg18

Reply:
I have mentioned a comment for flag that when range changes it is set to true. What more needs to be added ?